agent/proxycfg/snapshot.go

```diff
@@ -124,28 +168,38 @@ type ConfigSnapshot struct {
 
 	ServerSNIFn ServerSNIFunc
 	Roots       *structs.IndexedCARoots
+	// Used for connect-proxy and ingress-gateway proxies
+	Leaf *structs.IssuedCert
 
 	// connect-proxy specific
 	ConnectProxy configSnapshotConnectProxy
 
 	// mesh-gateway specific
 	MeshGateway configSnapshotMeshGateway
 
+	// ingress-gateway specific
+	IngressGateway configSnapshotIngressGateway
+
 	// Skip intentions for now as we don't push those down yet, just pre-warm them.
 }
 
 // Valid returns whether or not the snapshot has all required fields filled yet.
 func (s *ConfigSnapshot) Valid() bool {
 	switch s.Kind {
 	case structs.ServiceKindConnectProxy:
-		return s.Roots != nil && s.ConnectProxy.Leaf != nil
+		return s.Roots != nil && s.Leaf != nil
 	case structs.ServiceKindMeshGateway:
 		if s.ServiceMeta[structs.MetaWANFederationKey] == "1" {
 			if len(s.MeshGateway.ConsulServers) == 0 {
 				return false
 			}
 		}
 		return s.Roots != nil && (s.MeshGateway.WatchedServicesSet || len(s.MeshGateway.ServiceGroups) > 0)
+	case structs.ServiceKindIngressGateway:
+		return s.Roots != nil &&
+			s.Leaf != nil &&
+			s.IngressGateway.Config != nil &&
+			s.IngressGateway.ServiceLists != nil
```

Comment:
Should this be `len(s.IngressGateway.ServiceLists) > 0` instead? Or is it valid to have an empty service list and you just need to disambiguate between "not set yet" and "legitimately set to nothing"? If it's the latter then you could do the companion boolean approach like `WatchedServicesSet + ServiceGroups` works for mesh gateways.

Reply:
It's valid to have an empty list, we are using `nil` to disambiguate in this case. Adding something like a `ServiceListsSet` seems reasonable here to match `WatchedServicesSet `.

Reply:
Any time I treat a `nil` slice vs an empty slice differently I ultimately ending up regretting the decision somewhere along the way.

Reply:
We've removed this field from the ConfigSnapshot since we've decided to go with a different RPC mechanism that sends only the required set of services to watch.

Thus, we don't need to maintain a list of all services by namespace anymore.